### PR TITLE
Use setTimer on Nominator.setupTimer

### PR DESCRIPTION
**Pull request by TrustHenry**
`ITimer` interface has been newly created and this is implemented as `VibeTimer` and `LocalRestTimer`.
Currently, it only defined a stop function in the `ITimer`, which can be expanded.
Also, the `Nominator.setupTimer` uses `TaskManager`'s `setTimer` new task which waits until a timer expires and call

Relates #872

### CODE REVIEW

File: source/agora/consensus/protocol/Nominator.d [hunk 1]
```diff
@@ -526,15 +526,36 @@ extern(D):
...
+                    timer.stop();
+                    return;
+                }
+                callCPPDelegate(callback);
+            }
 
-            callCPPDelegate(callback);
+            callDelegate(); // avoid delay
+            timer = this.taskman.setTimer(timeout.msecs,
+                &callDelegate, Periodic.Yes);
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: This is wrong, it was not a periodic timer. There was no infinite loop.

File: source/agora/consensus/protocol/Nominator.d [hunk 2]
```diff
@@ -526,15 +526,36 @@ extern(D):
...
+                if (timer_id < this.active_timer_ids[type])
+                {
+                    timer.stop();
+                    return;
+                }
+                callCPPDelegate(callback);
+            }
 
-            callCPPDelegate(callback);
+            callDelegate(); // avoid delay
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: This is also wrong. Notice that there was a `wait` and then it was called. There is a specific reason for this. These are timeout routines.

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: I'll show you exactly how to do it.

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]

**Comment by AndrejMitrovic** [comment]
This is what `setupTimer` should look like:

```D
public override void setupTimer (ulong slot_idx, int timer_type,
    milliseconds timeout, CPPDelegate!SCPCallback* callback)
{
    scope (failure) assert(0);

    const type = cast(TimerType) timer_type;
    assert(type >= TimerType.min && type <= TimerType.max);
    if (callback is null || timeout == 0)
    {
        auto timer = this.active_timers[type];
        if (timer !is null)
        {
            timer.stop();
            this.active_timers[type] = null;
        }

        return;
    }

    this.active_timers[type] = this.taskman.setTimer(
        timeout.msecs, { callCPPDelegate(callback); });
}
```

**Comment by AndrejMitrovic** [comment]
And you should change `active_timers` variable to this:

```D
private ITimer[TimerType.max + 1] active_timers;
```

**Comment by AndrejMitrovic** [comment]
Also remove the `last_timer_id` variable.

**Comment by AndrejMitrovic** [comment]
@TrustHenry actually you will have to fix the implementation in your LocalRest PR. We need an ability to stop a timer after `sleep(timeout)` is called but before `dg()` is called.

**Comment by AndrejMitrovic** [comment]
The current tests will pass with the changes I proposed, however that is only because non of the nodes every time out during consensus (there is a lack of tests).

**Comment by TrustHenry** [comment]
Very sharp.
I also thought there was a problem with the timing.

**Comment by TrustHenry** [comment]
> The current tests will pass with the changes I proposed, however that is only because non of the nodes every time out during consensus (there is a lack of tests).

@AndrejMitrovic could you tell me more details about this?

**Comment by AndrejMitrovic** [comment]
> @AndrejMitrovic could you tell me more details about this?

This code might be a little hard to follow but let me explain:

https://github.com/bpfkorea/agora/blob/52e8aa0cfe7852954ce6430ab57cde0c153b2911/source/scpp/src/scp/BallotProtocol.cpp#L512-L515

The C++ SCP code calls the D code and asks it to set up a timer. If the timer has the chance to execute (the timeout expired and run() is called), then that run() function will call this:

https://github.com/bpfkorea/agora/blob/52e8aa0cfe7852954ce6430ab57cde0c153b2911/source/scpp/src/scp/BallotProtocol.cpp#L527-L531

But I'm not sure if this has been tested in Agora yet. It's worth looking into at some point.

**Comment by TrustHenry** [comment]
This requires a localRest update.
https://github.com/Geod24/localrest/pulls

File: source/agora/common/Task.d [hunk 3]
```diff
@@ -73,14 +73,80 @@ public class TaskManager
             dg = This delegate will be called when the timer fires.
             periodic = Specifies if the timer fires repeatedly or only once.
 
+        Returns:
+            A `ITimer` instance with the ability to control the VibeTimer
```
Comment (AndrejMitrovic) [inline source/agora/common/Task.d]:
```suggestion
            An `ITimer` interface with the ability to control the timer
```

File: source/agora/common/Task.d [hunk 4]
```diff
@@ -73,14 +73,80 @@ public class TaskManager
...
-
         static import vibe.core.core;
-        vibe.core.core.setTimer(timeout, dg, periodic);
+        return new VibeTimer(vibe.core.core.setTimer(timeout, dg, periodic));
+    }
+}
+
+/*******************************************************************************
+
+    Define the ability to control the timer
```
Comment (AndrejMitrovic) [inline source/agora/common/Task.d]:
```suggestion
    Defines an abstraction over the timer implementation
```

File: source/agora/common/Task.d [hunk 5]
```diff
@@ -73,14 +73,80 @@ public class TaskManager
...
+
+*******************************************************************************/
+
+public interface ITimer
+{
+    /***************************************************************************
+
+        Stop the timer executed by the setTimer
```
Comment (AndrejMitrovic) [inline source/agora/common/Task.d]:
```suggestion
        Stop or cancel this timer
```

File: source/agora/common/Task.d [hunk 6]
```diff
@@ -73,14 +73,80 @@ public class TaskManager
...
+
+    ***************************************************************************/
+
+    public void stop();
+}
+
+/*******************************************************************************
+
+    Vibe only timer
```
Comment (AndrejMitrovic) [inline source/agora/common/Task.d]:
```suggestion
    Vibe.d timer
```

File: source/agora/common/Task.d [hunk 7]
```diff
@@ -73,14 +73,80 @@ public class TaskManager
...
+
+*******************************************************************************/
+
+final class VibeTimer : ITimer
```
Comment (AndrejMitrovic) [inline source/agora/common/Task.d]:
```suggestion
private final class VibedTimer : ITimer
```

File: source/agora/common/Task.d [hunk 8]
```diff
@@ -73,14 +73,80 @@ public class TaskManager
...
+    }
+}
+
+/*******************************************************************************
+
+    LocalRest only timer (for unittests)
+
+*******************************************************************************/
+
+final class LocalRestTimer : ITimer
```
Comment (AndrejMitrovic) [inline source/agora/common/Task.d]: This doesn't belong here, but in `Base.d` where it's used. And it should be `private` too.

File: source/agora/consensus/protocol/Nominator.d [hunk 9]
```diff
@@ -74,11 +74,8 @@ public extern (C++) class Nominator : SCPDriver
     private alias TimerType = Slot.timerIDs;
     static assert(TimerType.max == 1);
 
-    /// Tracks unique incremental timer IDs
-    private ulong[TimerType.max + 1] last_timer_id;
-
-    /// Timer IDs with >= of the active timer will be allowed to run
-    private ulong[TimerType.max + 1] active_timer_ids;
+    /// The active timer will be allowed to run
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]:
```suggestion
    /// Currently active timers grouped by type
```

File: source/agora/consensus/protocol/Nominator.d [hunk 10]
```diff
@@ -520,24 +517,20 @@ extern(D):
...
-
-            // timer was cancelled
-            if (timer_id < this.active_timer_ids[type])
-                return;
-
-            callCPPDelegate(callback);
-        });
+        this.active_timers[type] = this.taskman.setTimer(
+            timeout.msecs, { callCPPDelegate(callback); });
+        }
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]:
```suggestion
```

was this a typo? It seems out of place.

File: source/agora/test/Base.d [hunk 11]
```diff
@@ -276,12 +276,16 @@ public class LocalRestTaskManager : TaskManager
             dg = This delegate will be called when the timer fires.
             periodic = Specifies if the timer fires repeatedly or only once.
 
+        Return:
+            A `LocalRsetTimer` instance with the ability to control the timer
```
Comment (AndrejMitrovic) [inline source/agora/test/Base.d]:
```suggestion
            An `ITimer` interface with the ability to control the timer
```

Also `LocalRsetTimer` was a typo. It would be good to catch these typos before requesting a review. An initial self-review is always good.

**Review by AndrejMitrovic — DISMISSED** [review]
Just small nits.

**Comment by TrustHenry** [comment]
@AndrejMitrovic Thank you for your review.
All suggestions have been applied.

File: source/agora/common/Task.d [hunk 12]
```diff
@@ -73,14 +73,56 @@ public class TaskManager
             dg = This delegate will be called when the timer fires.
             periodic = Specifies if the timer fires repeatedly or only once.
 
+        Returns:
+            An `ITimer` interface with the ability to control the timer
```
Comment (Geod24) [inline source/agora/common/Task.d:77]:
```suggestion
            An `ITimer` interface with the ability to control the timer

```

File: source/agora/common/Task.d [hunk 13]
```diff
@@ -73,14 +73,56 @@ public class TaskManager
...
+
+public interface ITimer
+{
+    /***************************************************************************
+
+        Stop or cancel this timer
+
+    ***************************************************************************/
+
+    public void stop();
```
Comment (Geod24) [inline source/agora/common/Task.d]: Vibe's Timer is `nothrow @safe`, but localrest is not annotated.
Also:
```suggestion
    public void stop ();
```

File: source/agora/common/Task.d [hunk 14]
```diff
@@ -73,14 +73,56 @@ public class TaskManager
...
+
+    private Vibe.Timer timer;
+
+    public this (Vibe.Timer timer)
+    {
+        this.timer = timer;
+    }
+
+    /// Ditto
+    public override void stop()
```
Comment (Geod24) [inline source/agora/common/Task.d]:
```suggestion
    public override void stop () @safe nothrow
```

File: source/agora/common/Task.d [hunk 15]
```diff
@@ -73,14 +73,56 @@ public class TaskManager
...
+
+*******************************************************************************/
+
+private final class VibedTimer : ITimer
+{
+    static import Vibe = vibe.core.core;
+
+    private Vibe.Timer timer;
+
+    public this (Vibe.Timer timer)
```
Comment (Geod24) [inline source/agora/common/Task.d]:
```suggestion
    public this (Vibe.Timer timer) @safe pure nothrow @nogc
```

File: source/agora/consensus/protocol/Nominator.d [hunk 16]
```diff
@@ -520,22 +517,18 @@ extern(D):
         assert(type >= TimerType.min && type <= TimerType.max);
         if (callback is null || timeout == 0)
         {
-            // signal deactivation of all current timers with this timer type
-            this.active_timer_ids[type] = this.last_timer_id[type] + 1;
+            auto timer = this.active_timers[type];
+            if (timer !is null)
```
Comment (Geod24) [inline source/agora/consensus/protocol/Nominator.d]: You can put those two statements in one:

```suggestion
            if (auto timer = this.active_timers[type])
```

File: source/agora/test/Base.d [hunk 17]
```diff
@@ -277,12 +277,40 @@ public class LocalRestTaskManager : TaskManager
...
+
+    public this (LocalRest.Timer timer)
+    {
+        this.timer = timer;
+    }
+
+    /// Ditto
+    public override void stop()
+    {
+        this.timer.stop();
```
Comment (Geod24) [inline source/agora/test/Base.d]: Same things about attributes and style

**Review by Geod24 — CHANGES_REQUESTED** [review]

File: source/agora/common/Task.d [hunk 18]
```diff
@@ -73,14 +73,57 @@ public class TaskManager
...
+
+*******************************************************************************/
+
+private final class VibedTimer : ITimer
+{
+    static import Vibe = vibe.core.core;
+
+    private Vibe.Timer timer;
+
+    public this (Vibe.Timer timer) @safe nothrow
```
Comment (Geod24) [inline source/agora/common/Task.d:119]: Missing `pure @nogc`

File: source/agora/test/Base.d [hunk 19]
```diff
@@ -277,12 +277,40 @@ public class LocalRestTaskManager : TaskManager
...
+
+*******************************************************************************/
+
+private final class LocalRestTimer : ITimer
+{
+    import LocalRest = geod24.LocalRest;
+
+    private LocalRest.Timer timer;
+
+    public this (LocalRest.Timer timer)
```
Comment (Geod24) [inline source/agora/test/Base.d:305]: Missing attributes

File: source/agora/test/Base.d [hunk 20]
```diff
@@ -277,12 +277,40 @@ public class LocalRestTaskManager : TaskManager
...
+    {
+        this.timer = timer;
+    }
+
+    /// Ditto
+    public override void stop ()
```
Comment (Geod24) [inline source/agora/test/Base.d:311]: We can proceed as is, or you can make a PR to LocalRest to fix the attributes. I'll merge it fast.

File: source/agora/common/Task.d (hunk 18 above)
Comment (TrustHenry) [reply source/agora/common/Task.d:119]:
```
source/agora/common/Task.d(119,12): Error: pure constructor agora.common.Task.VibedTimer.this cannot call impure destructor vibe.core.core.Timer.~this
source/agora/common/Task.d(119,12): Error: @nogc constructor agora.common.Task.VibedTimer.this cannot call non-@nogc destructor vibe.core.core.Timer.~this
source/agora/common/Task.d(121,20): Error: pure constructor agora.common.Task.VibedTimer.this cannot call impure function vibe.core.core.Timer.__postblit
source/agora/common/Task.d(121,20): Error: @nogc constructor agora.common.Task.VibedTimer.this cannot call non-@nogc function vibe.core.core.Timer.__postblit
```

Comment (Geod24) [reply source/agora/common/Task.d:119]: Oh man. LGTM then.

File: source/agora/test/Base.d (hunk 20 above)
Comment (TrustHenry) [reply source/agora/test/Base.d:311]: https://github.com/Geod24/localrest/blob/2260b8eae7daec3d1991a1859bec759cc729a653/source/geod24/concurrency.d#L945
The attributes are not safe. Is it intended?

**Review by Geod24 — APPROVED** [review]

File: source/agora/consensus/protocol/Nominator.d [hunk 21]
```diff
@@ -520,22 +517,16 @@ extern(D):
...
-            this.taskman.wait(timeout.msecs);
-
-            // timer was cancelled
-            if (timer_id < this.active_timer_ids[type])
-                return;
-
-            callCPPDelegate(callback);
-        });
+        this.active_timers[type] = this.taskman.setTimer(
+            timeout.msecs, { callCPPDelegate(callback); });
```
Comment (Geod24) [inline source/agora/consensus/protocol/Nominator.d:527]: Looks like previously set timers aren't stopped ?

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d:527]: @TrustHenry could you also check if `this.active_timers[type]` exists here, and if it does call stop on it?

Comment (TrustHenry) [reply source/agora/consensus/protocol/Nominator.d:527]: In the unit test, Timer.stops confirmed a total of 106 occurrences.
This is also `if (auto timer = this.active_timers[type])`

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d:527]: I meant that we should probably do something like this:

```diff
diff --git a/source/agora/consensus/protocol/Nominator.d b/source/agora/consensus/protocol/Nominator.d
index 4a0f115e2..381c6faca 100644
--- a/source/agora/consensus/protocol/Nominator.d
+++ b/source/agora/consensus/protocol/Nominator.d
@@ -515,16 +515,15 @@ extern(D):

         const type = cast(TimerType) timer_type;
         assert(type >= TimerType.min && type <= TimerType.max);
-        if (callback is null || timeout == 0)
+        if (auto timer = this.active_timers[type])
         {
-            if (auto timer = this.active_timers[type])
-            {
-                timer.stop();
-                this.active_timers[type] = null;
-            }
-            return;
+            timer.stop();
+            this.active_timers[type] = null;
         }

+        if (callback is null || timeout == 0)
+            return;
+
         this.active_timers[type] = this.taskman.setTimer(
             timeout.msecs, { callCPPDelegate(callback); });
     }
```

So always stop any existing timers. And then if callback is null we just return, otherwise we set a timer.

Comment (TrustHenry) [reply source/agora/consensus/protocol/Nominator.d:527]: Was there a leak here?
The Code of @AndrejMitrovic  applied successfully.

Comment (TrustHenry) [reply source/agora/consensus/protocol/Nominator.d:527]: If this callback is not null?

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d:527]: It's possible.

**Review by Geod24 — APPROVED** [review]
